src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+        if (self.pos + lookahead < self.buf.len) {
+            const ch = self.buf[self.pos + lookahead];
+            return ch >= start and ch <= end;
+        }
+        return false;
+    }
+
+    /// Requires that the next byte is `ch`, otherwise pushes a message.
+    pub fn require(self: *Cursor, ch: u8, kind: MessageKind) void {
+        if (self.peek() == ch) {
+            self.pos += 1;
+        } else {
+            self.pushMessageHere(kind);
+        }
+    }
+
+    /// Chomps “trivia” (whitespace, comments, etc.) and returns an optional indent.
+    pub fn chompTrivia(self: *Cursor) ?Indent {
+        var sawNewline = false;
+        var indent = Indent.init();
+
+        while (self.pos < self.buf.len) {
+            const b = self.buf[self.pos];
+            if (b == ' ') {
+                self.pos += 1;
+                if (sawNewline) indent.num_spaces += 1;
+            } else if (b == '\t') {
+                self.pos += 1;
+                if (sawNewline) indent.num_tabs += 1;
+            } else if (b == '\n') {
+                self.pos += 1;
+                sawNewline = true;
+                indent = Indent.init();
+            } else if (b == '\r') {
+                self.pos += 1;
+                sawNewline = true;
+                indent = Indent.init();
+                if (self.pos < self.buf.len and self.buf[self.pos] == '\n') {
+                    self.pos += 1;
+                } else {
+                    self.pushMessageHere(.MisplacedCarriageReturn);
+                }
+            } else if (b == '#') {
+                self.pos += 1;
+                while (self.pos < self.buf.len and self.buf[self.pos] != '\n') {
+                    self.pos += 1;
+                }
+            } else if (b >= 0 and b <= 31) {
+                self.pushMessageHere(.AsciiControl);
+                self.pos += 1;
+            } else {
+                break;
+            }
+        }
+        if (sawNewline) return indent;
+        return null;
+    }
+
+    fn maybeMessageForUppercaseBase(self: *Cursor, b: u8) void {
+        if (b == 'X' or b == 'O' or b == 'B') {
+            self.pushMessageHere(.UppercaseBase);
+        }
+    }
+
+    pub fn peekUnicode(self: *Cursor) UnicodeInfo {
+        const len3 = std.unicode.utf8ByteSequenceLength(self.buf[self.pos]) catch {
+            return .{ .kind = .Invalid, .length = 1 };
+        };
+        const len: usize = @intCast(len3);
+        const utf8_char = std.unicode.utf8Decode(self.buf[self.pos .. self.pos + len]) catch {
+            return .{ .kind = .Invalid, .length = 1 };
+        };
+        switch (self.gc.gc(utf8_char)) {
+            .Lu, .Lt => return .{ .kind = .LetterUpper, .length = len },
+            .Ll, .Lm, .Lo => return .{ .kind = .LetterNotUpper, .length = len },
+            .Nd, .Nl, .No => return .{ .kind = .Digit, .length = len },
+            else => return .{ .kind = .Other, .length = len },
+        }
```

Comment:
Is this the only use of gencat? I wonder if we need a library for this at all. To be fair, I'm not sure I understand why unicode would result in a letter or digit. Isn't that just normal ascii? Like can't we check for that manually pretty easily?

Reply:
Letter covers other languages as well. The current rust parser uses `char::is_uppercase` and `char::is_lowercase` which are both unicode aware, and allow things like Θ (uppercase theta) and θ (lowercase theta). As well as accented characters, etc.

I generally agree on trying to not use this rather bulky library. Eventually I want to just hard-code the (hopefully small) set of specific ranges we want to allow, but this is simpler for now.